architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,302 @@
+
+### Only the Model is Modified
+
+Since the [`SFTTrainer`] is designed to work with generic pytorch models, modifying the model is much less intrusive to the training pipeline, then say, modifying [`SFTTrainer`] itself. The hope is then if we constrain ourselves to modify only the model, that we can implement all the method plugins (e.g., quantization, distributed training, etc) that we hope for. 
+
+The framework is designed to only modify them model at two integration points in `sft_trainer.py`. The primary motivation for this is easy code maintenance:
+1. an *optional* `loading` method that acts as a drop-in replacement for `AutoModel.from_pretrained`. 
+2. an *optional* `agumentation` method that provides a way to perform *minor* adjustments to an already instantiated model 
+3. an *optional* `callback` method to install `TrainerCallbacks` (if needed, e.g. custom save logic).
+
+```python
+class TuningAccelerationPlugin:
+
+    # if specified, will restricted plugin to specified model archs
+    # - useful if method is restricted to certain model architectures, e.g., only used 
+    #  for MoEs
+    restricted_model_archs: Set = None
+
+    # if specified, will check if the package/s is/are installed
+    required_packages: Set = None
+
+    # if True, will check if environment supports CUDA Toolkit
+    require_cuda_tools: bool = False
+
+    def loading(model_path: str, **kwargs):
+        pass
+
+    # augment model or accelerator object
+    def augmentation(model: nn.Module, **kwargs):
+        # accelerator = kwargs.get('accelerator') # if needed for configs (e.g. FSDP)
+        pass
+
+    def callbacks(model: nn.Module, **kwargs):
+        cbks = []
+        return cbks
+```
+
+Even though they are all optional, at least one out of the three should be implemented.
+
+### Dependency Management
+
+Take note:
+- all plugin deps must be enforced to be optional deps in `pyproject.toml`, see [116](https://github.com/foundation-model-stack/fms-hf-tuning/pull/116). If the dep is not installed, and the plugin is enabled, raise exception.
+- any plugin that requires CUDA build tools (e.g. `triton` kernels) will need to be run in with [CUDA Toolkit dependencies (see this link for an example of a Debian installation)](https://developer.nvidia.com/cuda-12-2-0-download-archive?target_os=Linux&target_arch=x86_64&Distribution=Debian&target_version=11&target_type=deb_local). 
+    * whenever CUDA is needed, the framework will check for the CUDA_TOOLS dependency.
+
+### Minimal and Controlled Changes to Training Script
+
+All proposed code changes to `sft_trainer.py` contained in minimal lines of code:
+- Plugins loaded by discovery; transparent to `sft_trainer.py`.
+- Plugin configuration automatically parsed.
+- Passthrough to original operation if `Framework` is disabled.
+
+```python
+from tuning.acceleration import AccelerationFramework
```

Comment:
things I am not very clear on:
1. How does example above `TuningAccelerationPlugin:` relate here ?It is not even imported in this example, so is it not used here?  Can we get one complete example in ADR? What does `AccelerationFramework` contain? 

2. what happens when there are multiple frameworks all imported in main code? like what if theer is one for quantization and one for something else? at train time what flag will determine which framework to use? Here it seems very hard coded to AccelartionFramework `if framework_args.config_file is not None:
    framework = AccelerationFramework(framework_args.config_file)`
  
Will there not be more than one?

 3. I think I may not be understanding differenec between Framework and plugin in this ADR. From what I understood is plugin is the class that has custom loaders/augmentors . What is Framework? Is Framework the yaml file + Plugin? I see it has the flags `framework.requires_custom_loading:` , but it also has the loaders `framework.model_loader ` 
 
 4. did you mean `framework.loading()` or `framework.model_loader` in line 139 ? Should framework have a model_loader property?

Reply:
@Ssukriti thanks for your careful reading. Have attempted to address in https://github.com/foundation-model-stack/fms-hf-tuning/pull/119/commits/da8c9f8905cd50bcf9d59b8590f39b86cb5c5bdf

Just take note in the rewrite there are renames. I shall refer to them with the new names
- `TuningAccelerationPlugin` -> `AccelerationPlugin`
- `Framework` -> `AccelerationFramework`

> How does example above TuningAccelerationPlugin: relate here ?It is not even imported in this example, so is it not used here? Can we get one complete example in ADR? What does AccelerationFramework contain?

I take it that you mean `AccelerationPlugin` was not imported `sft_trainer.py` example. Yes it was omission on my part. I have now included:
- (almost) full implementation of `AccelerationFramework`. See new section **Implmentation of Acceleration Framework Class**. This is the thing that is imported in `sft_trainer.py` and it will construct and manage all the `AccelerationPlugin`. This management is *transparent* to `sft_trainer.py`, that is why it does not appear. 
- Also in **AccelerationPlugin Base Class**, I have further fleshed out `AccelerationPlugin`, and show that all implemented plugins need to be registered to `PLUGIN_REGISTRATIONS`. That is how `AccelerationFramework` will know which plugins are registered and available for loading.
- In section **Detailed Design**.  this shows a concrete implemntation of `AutoGPTQAccelerationPlugin` that inherits from `AccelerationPlugin`. And the end of the class definition, you can see we make a `register_plugin` call to register the plugin. When registering, we leave a bread crumb to inform `AccelerationFramework` that if you see this key in the configuration file, then promote the plugin to active status.
    ```
    configuration_paths=["peft.quantization.auto_gptq"], 
    ```
- In the following subsection *Configuration To Load AutoGPTQ LoRA Plugin**, I show an example of a configuraiton file used to activate `AutoGPTQAccelerationPlugin`

Reply:
> what happens when there are multiple frameworks all imported in main code? like what if theer is one for quantization and one for something else? at train time what flag will determine which framework to use? Here it seems very hard coded to AccelartionFramework if framework_args.config_file is not None: framework = AccelerationFramework(framework_args.config_file)

@Ssukriti No I intend `AccelartionFramework` to be a *singleton*. On referring to the newly updated `sft_trainer.py`, `AccelartionFramework` will be created only if `AccelerationFrameworkArguments.acceleration_framework_config_file` is specified.
- plugins are transparent to user. The `AccelartionFramework` hides logic for managing and loading plugins to create a simple interfacing point.
- Inside that config file, we could have multiple plugin descriptions, that loads multiple plugins. I give another example below, to show something we have in our `dev` code to that will load 2 plugins, the second plugin installs fast kernels on top of the first plugin that loads an auto-gptq (LoRA) trainable model.
  ```yaml
  plugins:
  
    # PEFT-related acceleration
    peft:
  
      # quantization-releated acceleration
      # e.g., kernels for quantized base weights
      quantization: 
  
        # AutoGPTQ quantized base weights.
        auto_gptq:
          kernel: triton_v2
          from_quantized: True
  
        unsloth:
          stackable: 
            base_layer: auto_gptq # bitsandbytes
            fused_lora: True
            fast_loss: True
            fast_rsm_layernorm: True
            fast_rope_embeddings: True
  ```

Reply:
> I think I may not be understanding differenec between Framework and plugin in this ADR. From what I understood is plugin is the class that has custom loaders/augmentors . What is Framework? Is Framework the yaml file + Plugin? I see it has the flags framework.requires_custom_loading: , but it also has the loaders framework.model_loader 

Sorry again for the confusion, the user only interfaces with `AccelerationFramework`. The `AccelerationPlugin` also has `model_loader` and `augmentation` functinos, but we rely on the `AccelerationFramework` to call them; the user does not call them directly

I hope that the full implementation of `AccelerationFramework` provided under the new section **Implmentation of Acceleration Framework Class** can help clarify what im saying here

Reply:
> did you mean framework.loading() or framework.model_loader in line 139 ? Should framework have a model_loader property?

Oh, i did mean it as is. If i reprise the code here:

```
# Minor Change 2: custom loader (if necessary)
_model_loader = AutoModelForCausalLM.from_pretrained # default
if framework is not None and framework.requires_custom_loading:
    _model_loader = framework.model_loader # drop in replacement

# will passthrough the default loader if framework is disabled
model = _model_loader(
    model_args.model_name_or_path,
    cache_dir=train_args.cache_dir,
    torch_dtype=get_torch_dtype(model_args.torch_dtype),
    attn_implementation="flash_attention_2" if model_args.use_flash_attn else None,
)
```

- if `framework is not None and framework.requires_custom_loading` is False, then 
  ```
  _model_loader = AutoModelForCausalLM.from_pretrained 
  ```
  which means that on the line `model = _model_loader(...)` then `model` will be loaded by `AutoModelForCausalLM.from_pretrained`.
- similarly otherwise then `model` will be loaded by `framework.model_loader `, which is a bona-fide function, and internally has logic to call the plugin's model_loader to load the model. See the full implementation of `AccelerationFramework` that is now provided in https://github.com/foundation-model-stack/fms-hf-tuning/pull/119#:~:text=%E2%80%A6-,da8c9f8,-fabianlim%20force%2Dpushed